src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs

```diff
@@ -84,9 +132,89 @@ internal Diagnostic GetErrorAndMissingAssemblyIdentities(DiagnosticBag diagnosti
                 missingAssemblyIdentities = ImmutableArray<AssemblyIdentity>.Empty;
             }
 
-            return diagnosticsEnumerable.FirstOrDefault(d => d.Severity == DiagnosticSeverity.Error);
+            useReferencedModulesOnly = diagnosticsEnumerable.All(HasDuplicateTypesOrAssemblies);
+
+            var firstError = diagnosticsEnumerable.FirstOrDefault();
+            Debug.Assert(firstError != null);
+
+            var simpleMessage = firstError as SimpleMessageDiagnostic;
+            return (simpleMessage != null) ?
+                simpleMessage.Message :
+                formatter.Format(firstError, preferredUICulture ?? CultureInfo.CurrentUICulture);
         }
 
+        internal abstract bool HasDuplicateTypesOrAssemblies(Diagnostic diagnostic);
+
         internal abstract ImmutableArray<AssemblyIdentity> GetMissingAssemblyIdentities(Diagnostic diagnostic);
+
+        protected sealed class SimpleMessageDiagnostic : Diagnostic
```

Comment:
What does one of these look like in the debugger?  Would it be worthwhile to add a DebuggerDisplayAttribute?